code/controllers/subsystem/spatial_gridmap.dm

```diff
@@ -354,27 +369,32 @@ SUBSYSTEM_DEF(spatial_grid)
 	if(QDELETED(new_target))
 		CRASH("qdeleted or null target trying to enter the spatial grid!")
 
-	if(!target_turf || !new_target.important_recursive_contents)
-		CRASH("null turf loc or a new_target without important_recursive_contents trying to enter the spatial grid!")
+	if(!target_turf || !(new_target.important_recursive_contents || new_target.flags_1 & SPATIAL_GRID_MANAGED_1))
+		CRASH("null turf loc or a new_target that doesn't support it trying to enter the spatial grid!")
 
-	var/x_index = ROUND_UP(target_turf.x / SPATIAL_GRID_CELLSIZE)
-	var/y_index = ROUND_UP(target_turf.y / SPATIAL_GRID_CELLSIZE)
+	var/x_index = GET_SPATIAL_INDEX(target_turf.x)
+	var/y_index = GET_SPATIAL_INDEX(target_turf.y)
 	var/z_index = target_turf.z
 
 	var/datum/spatial_grid_cell/intersecting_cell = grids_by_z_level[z_index][y_index][x_index]
+	if(new_target.important_recursive_contents)
+		if(new_target.important_recursive_contents[RECURSIVE_CONTENTS_CLIENT_MOBS])
+			GRID_CELL_SET(intersecting_cell.client_contents, new_target.important_recursive_contents[SPATIAL_GRID_CONTENTS_TYPE_CLIENTS])
 
-	if(new_target.important_recursive_contents[RECURSIVE_CONTENTS_CLIENT_MOBS])
-		GRID_CELL_SET(intersecting_cell.client_contents, new_target.important_recursive_contents[SPATIAL_GRID_CONTENTS_TYPE_CLIENTS])
+			SEND_SIGNAL(intersecting_cell, SPATIAL_GRID_CELL_ENTERED(RECURSIVE_CONTENTS_CLIENT_MOBS), new_target)
 
-		SEND_SIGNAL(intersecting_cell, SPATIAL_GRID_CELL_ENTERED(RECURSIVE_CONTENTS_CLIENT_MOBS), new_target)
+		if(new_target.important_recursive_contents[RECURSIVE_CONTENTS_HEARING_SENSITIVE])
+			GRID_CELL_SET(intersecting_cell.hearing_contents, new_target.important_recursive_contents[RECURSIVE_CONTENTS_HEARING_SENSITIVE])
 
-	if(new_target.important_recursive_contents[RECURSIVE_CONTENTS_HEARING_SENSITIVE])
-		GRID_CELL_SET(intersecting_cell.hearing_contents, new_target.important_recursive_contents[RECURSIVE_CONTENTS_HEARING_SENSITIVE])
+			SEND_SIGNAL(intersecting_cell, SPATIAL_GRID_CELL_ENTERED(RECURSIVE_CONTENTS_HEARING_SENSITIVE), new_target)
 
-		SEND_SIGNAL(intersecting_cell, SPATIAL_GRID_CELL_ENTERED(RECURSIVE_CONTENTS_HEARING_SENSITIVE), new_target)
+	if(istype(new_target, /obj/machinery/atmospherics))
```

Comment:
i really dont like this. its O(1) now at least but still. this makes me lean towards wanting "x is in the grid channels a, b, and c" to be turned into a trait with several sources but oh god list memory costs

Reply:
You could use a list for this, maybe a hash -> list global list so it's cheap to store.
Sources gets harder, would need a good reason to want that.

Reply:
34b7b261467383136cc92f65f35f4642e9ef43f9